pkg/controller/pkg/revision/reconciler.go

```diff
@@ -0,0 +1,565 @@
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	pr := r.newPackageRevision()
+	if err := r.client.Get(ctx, req.NamespacedName, pr); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		log.Debug(errGetPackageRevision, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPackageRevision)
+	}
+
+	reader, err := r.backend.Init(ctx, parser.PodClient(r.kubeClient), parser.PodName(pr.GetInstallPod().Name))
+	if err != nil {
+		log.Debug(errInitParserBackend, "error", err)
+		pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+		return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+	pkg, err := r.parser.Parse(ctx, reader)
+	if err != nil {
+		log.Debug(errParseLogs, "error", err)
+		pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+
+	desiredState := pr.GetDesiredState()
+	pkgProvider, err := getProviderFromPkg(pkg)
+	if err != nil {
+		log.Debug(errPackageMeta, "error", err)
+		pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+		return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+
+	// If the parsed package metadata does not match the appropriate package
+	// revision, we don't want to reconcile it.
+	if pkgProvider != nil && pr.GetObjectKind().GroupVersionKind() != v1alpha1.ProviderRevisionGroupVersionKind {
+		log.Debug(errPackageTypeMismatch, "error", err)
+		pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+		return reconcile.Result{}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+	}
+
+	// If revision is inactive and is a provider, we must make sure controller
+	// is not running before we clean up resources.
+	if pkgProvider != nil && desiredState == v1alpha1.PackageRevisionInactive {
+		s, d := buildProviderDeployment(pkgProvider, pr, r.namespace)
+		if err := r.client.Delete(ctx, d); err != nil {
+			log.Debug(errDeleteProviderDeployment, "error", err)
+			pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+		if err := r.client.Delete(ctx, s); err != nil {
+			log.Debug(errDeleteProviderSA, "error", err)
+			pr.SetConditions(runtimev1alpha1.Unavailable(), runtimev1alpha1.ReconcileSuccess())
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
+		}
+	}
```

Comment:
I forget whether we discussed during design what would happen to managed resources that are 'orphaned' when a provider revision becomes inactive. I recall that their CRDs and CRs would continue to exist and be owned by an inactive package revision, but I suspect we'd be unable to delete any dangling CRs without a controller running to reconcile them since nothing would remove the finalizers.

Reply:
So as long as the revision exists, its objects will also exist (the inactive revision is now an Owner instead of Controller). You are correct that in order to clean up CRs after that, you would need to manually remove the finalizer. However, you could also just delete the revision, in which case the CRD would be GC'ed by k8s (no more owner refs), and all CRs for that CRD will also be cleaned up.

I think this is a good pattern as it makes use of kubernetes inherent functionality, rather than adding a bunch of logic on top. The one case where this could be difficult is when you update a package and then want to go and delete some but not all of the CRs for CRDs only owned by the old package revision (this would be a strange case because you would be electing to keep around CRs that were not being reconciled). In this case, I think we should be okay saying that it takes some manual work, and we could even add some utilities to `crank` to help with removing finalizers for things owned by an inactive package revision.

Reply:
I agree with the approach you propose, but are you sure that deleting a CRD will delete the CRs it defines, even if they have finalizers that never get removed?

Reply:
You are correct, the CRD will hang around as long as the CR still has a finalizer, so folks would need to go back and manually remove that finalizer (ref: https://github.com/kubernetes/kubernetes/issues/60538#issuecomment-680279481). This is not straightforward for us to address though. If you leave behind CRs when deleting a Revision (or its parent package), I'm not sure we can assume that you want us to remove arbitrary finalizers that the provider controller may have added to the CRs, and I definitely don't think we should ever explicitly delete CRs. Do you have any thoughts here?

Reply:
I really like this idea. I'll get it written up in an issue for tracking :+1:

Reply:
Opened https://github.com/crossplane/crossplane/issues/1732 to track